craft_parts/state_manager/reports.py

```diff
@@ -0,0 +1,61 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2018-2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Provide a report on why a step is outdated."""
+
+from craft_parts.steps import Step
+from craft_parts.utils import formatting_utils
+
+
+class OutdatedReport:
```

Comment:
With its usage, would this make more sense as an exception?

Reply:
I considered it some time ago but given how it's used during the lifecycle processing I prefer the current form (the lifecycle logic is not exception-driven, and handling dirty/outdated steps as exceptions didn't feel right)